go.mod

```diff
@@ -1036,6 +1073,19 @@ retract v0.37.0 // Contains dependencies on v0.36.0 components, which should hav
 // see https://github.com/distribution/distribution/issues/3590
 exclude github.com/docker/distribution v2.8.0+incompatible
 
-// some dependencies attempt to bring something like v1.8.2-0.20220303173753-edfe657b5405, which is older than v0.38.0
-// at the time of this inclusion, v0.38.0 was the latest version available (also tagged as v2.38.0)
-replace github.com/prometheus/prometheus => github.com/prometheus/prometheus v0.38.0
+// some dependencies attempt to bring something like v1.8.2-0.20220303173753-edfe657b5405, which is older than v0.39.1
+// at the time of this inclusion, v0.39.1 was the latest version available (also tagged as v2.39.1)
+replace github.com/prometheus/prometheus => github.com/prometheus/prometheus v0.39.1
+
+// loki needs grafana fork to be used
+replace github.com/hpcloud/tail => github.com/grafana/tail v0.0.0-20220426200921-98e8eb28ea4c
+
+// using fork for support LogpullReceived and LogpullFields for loki - see https://github.com/cloudflare/cloudflare-go/pull/743
+replace github.com/cloudflare/cloudflare-go => github.com/cyriltovena/cloudflare-go v0.27.1-0.20211118103540-ff77400bcb93
+
+// using fork for custom dialer support - see https://github.com/bradfitz/gomemcache/pull/86
+replace github.com/bradfitz/gomemcache => github.com/grafana/gomemcache v0.0.0-20220812141859-1e3ae89e91a7
+
+// fork containing a line-buffered logger which should improve logging performance.
+// TODO: submit PR to upstream and remove this
```

Comment:
I am worried about the large number of `replace` directives used here. `replace` directives hinder usability of the Opentelemetry Collector Builder since the user needs to understand what replace directives are needed for including a given component. They also make life harder for non-official distros, which run into basically the same problem.

My suggestion here would be to instead require these forks directly. This may need changes on the dependencies you use, but it's the only way to avoid this for builder users

Reply:
I made commit messages more clear, so it is more visible for users which `replace` is needed for loki.
I went through all of them and tried to get rid of some of them as much as possible, sent PR upstream, and discussed them with loki team. 
I understand that replaces are not convenient, but it is difficult to get rid of them right now. 
Requiring forks instead of upstream is not a clean solution for loki, since it will lead to a lot of changes in loki repo itself and in forks as well. Also, it will take more effort to switch back to the upstream in the future and will discourage from creating PRs to the upstream.

I would leave these replaces. And since I'm one of the owners of components that depend on loki, I will keep an eye on these replace and try to remove them

Reply:
Since the problems are scoped to builder users that want to use the `promtail` receiver, I will leave this up to @jpkrohling

Reply:
I suggest the following compromise: we'll commit to reaching out to the Loki team and expose this problem, working with them on a solution that suits us all. In the meantime, I'd suggest having the minimal set of replaces required to get this PR into a working state.

Reply:
Okay, sounds reasonable to me. I think we also need guidelines on replace directives more generally, I will look into that from my side :) Please post the tracking issue here once created so that future readers can follow the conversation

Reply:
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/16453